recommenders/evaluation/spark_evaluation.py

```diff
@@ -630,6 +659,43 @@ def _get_cosine_similarity(self, n_partitions=200):
             )
         return self.df_cosine_similarity
 
+    @staticmethod
+    @udf(returnType=DoubleType())
+    def sim_cos(v1, v2):
+        try:
+            p = 2
+            return float(v1.dot(v2)) / float(v1.norm(p) * v2.norm(p))
+        except AssertionError:
+            raise Exception(
+                "Dimension mismatch! The size of two input vectors should be the same."
+            )
+        except:
+            return 0
+
+    def _get_item_feature_similarity(self, n_partitions):
+        """Cosine similarity metric based on item feature vectors
+
+        The item indexes in the result are such that i1 <= i2.
+        """
+        if self.df_cosine_similarity is None:
+            self.df_cosine_similarity = (
+                self.item_feature_df.select(
+                    F.col(self.col_item).alias("i1"),
+                    F.col(self.col_item_features).alias("f1"),
+                )
+                .join(
+                    self.item_feature_df.select(
+                        F.col(self.col_item).alias("i2"),
+                        F.col(self.col_item_features).alias("f2"),
+                    ),
+                    (F.col("i1") < F.col("i2")),
+                )
+                .select("i1", "i2", self.sim_cos("f1", "f2").alias("sim"))
```

Comment:
This function uses the udf, whereas `_get_cooccurrence_similarity()` uses SQL operations. At the core, both functions compute the same thing (cosine similarity) but in different ways. It probably makes sense to be consistent between the two i.e. to use the most efficient approach in both. Also the computation could then be factored out to remove duplication.

Reply:
For function "_get_item_feature_similarity" I have to use a udf. I have not found a good way to combine both functions.

Reply:
Yes, it would be a bit more complicated. If you require that the features are sparse vectors (as in the example notebook), an alternative approach would be to first put them in tabular form inside a dataframe and then use SQL operations (similar to the co-occurrence case but with an additional SQL multiplication). I was curious whether that would be faster or slower than the UDF. Let's leave it for another occasion, after all the features could be dense in certain scenarios.